src/passes/OptimizeInstructions.cpp

```diff
@@ -1987,10 +1995,25 @@ struct OptimizeInstructions
       return;
     }
 
-    // The cast will not definitely succeed, but perhaps the heap type part of
-    // the cast will, at least. That would leave only nullability as an issue,
+    // The cast will not definitely succeed nor will it definitely fail.
+    //
+    // Perhaps the heap type part of the cast can be reasoned about, at least.
+    // E.g. if the heap type part of the cast is definitely compatible, but the
+    // cast as a whole is not, that would leave only nullability as an issue,
     // that is, this means that the input ref is nullable but we are casting to
     // non-null.
+    //
+    // Note that we could do something similar for a failed cast, that is,
+    // handle the situation where the entire cast might succeed, but the heap
+    // type part will definitely fail. For example, the input might be a
+    // nullable array while the output might be a nullable struct. That is, a
+    // situation where the only way the cast succeeds is if the input is null.
+    // However, optimizing this would mean emitting something like
+    //
+    //   ref == null ? null : trap
```

Comment:
It would be nice to update this comment to mention the `ref.cast null none` idea, which would be strictly better than this code pattern.

Reply:
But no need if this comment is going to be short-lived anyway.

Reply:
It's already deleted in the followup :smile: